File: leetcode/python/576_out_of_boundary_paths.py
class Solution:
    def __init__(self) -> None:
        self.MAX = 10**9 + 7

    def findPaths(self, m: int, n: int, maxMove: int, startRow: int, startColumn: int) -> int:
        dp = [[0] * n for _ in range(m)]
        dp[startRow][startColumn] = 1
        count: int = 0
        for _ in range(maxMove):
            temp = [[0] * n for _ in range(m)]
            for i in range(m):
                for j in range(n):
                    if i == 0:
                        count = (count + dp[i][j]) % self.MAX
                    if i == m - 1:
                        count = (count + dp[i][j]) % self.MAX
                    if j == 0:
                        count = (count + dp[i][j]) % self.MAX
                    if j == n - 1:
                        count = (count + dp[i][j]) % self.MAX

                    temp[i][j] = (dp[i-1][j] if i > 0 else 0) + \
                        (dp[i+1][j] if i < m-1 else 0) + \
                        (dp[i][j-1] if j > 0 else 0) + \
                        (dp[i][j+1] if j < n-1 else 0)
            dp = temp
        return count
                    
s = Solution()
print(s.findPaths(m = 2, n = 2, maxMove = 2, startRow = 0, startColumn = 0))
print(s.findPaths(m = 1, n = 3, maxMove = 3, startRow = 0, startColumn = 1))

File: leetcode/rust/src/576_out_of_boundary_paths.rs
impl Solution {
    pub fn find_paths(m: i32, n: i32, max_move: i32, start_row: i32, start_column: i32) -> i32 {
        let mut dp = vec![vec![vec![-1; n as usize]; m as usize]; max_move as usize + 1];
        
        fn find_path(m: i32, n: i32, max_move: i32, i: i32, j: i32, dp: &mut Vec<Vec<Vec<i32>>>) -> i32 {
            if i < 0 || i >= m || j < 0 || j >= n {
                return 1;
            }
            if max_move == 0 {
                return 0;
            }
            if dp[max_move as usize][i as usize][j as usize] != -1 {
                return dp[max_move as usize][i as usize][j as usize];
            }
            let mut res = 0;
            let MOD = 1_000_000_007;
            res = (res + find_path(m, n, max_move - 1, i - 1, j, dp)) % MOD;
            res = (res + find_path(m, n, max_move - 1, i + 1, j, dp)) % MOD;
            res = (res + find_path(m, n, max_move - 1, i, j - 1, dp)) % MOD;
            res = (res + find_path(m, n, max_move - 1, i, j + 1, dp)) % MOD;
            dp[max_move as usize][i as usize][j as usize] = res;
            res
        }

        find_path(m, n, max_move, start_row, start_column, &mut dp)
    }
}